apps/core/src/migration/service/convert-monitor-to-app-definition.ts

```diff
@@ -316,6 +331,81 @@ const convertProperties = (
   return getStaticProperties(widgetType);
 };
 
+const convertKpiOrTimelineWidget = (
```

Comment:
this function is doing a lot of calculation for coordinates and it is getting complicated with the many parts and cases involved.
1. Can you document the behaviors/cases?
2. Can we do something to encapsulate the calculation logic? e.g. `Generator` to help with traversing each metric

Reply:
That's fair, I just pushed an update to make this more clear

1. Added documentation to explain the logic more
2. Broke the function down more to make it clear and did some refactoring